### PR TITLE
Enable migrations from older database schemas.

### PR DESCRIPTION
**Pull request by evetion**
Fixes https://github.com/Deltares/Ribasim-NL/issues/138

This has no schema version (stored in the geopackage) yet, but this is something we could implement in the future.

### CODE REVIEW

**Comment by deltamarnix** [comment]
I would like to opt to add a geopackage version in this PR and let's call everything that doesn't have it version 0, and from now on we will write a `modelversion: 1`. Otherwise we might get conflicts when we figure out that more backwards compatibility is required in the coming months. Then while writing migration, first check if the version is correct.

- [ ] change QGIS writing
- [ ] change Ribasim python writing
- [ ] Write the geopackage version during migration if it was successful? (Maybe not, since it's only in memory migration)
- [ ] Check in your migration code what the version is to migrate

For now, let's keep it as simple as you have it, although I would rather see a pipeline that migrates step by step. But that's for another moment in time.

**Comment by visr** [comment]
This doesn't yet include removing `listen_node_type` columns:

```
ValidationError: 1 validation error for TableModel[DiscreteControlVariableSchema]
df
  Value error, Unrecognized column 'listen_node_type'.
```

**Comment by evetion** [comment]
@deltamarnix This now writes the schema version to the geopackage (also from QGIS), which is used to check whether to migrate or not. @visr This now also migrates the other tables with `listen_node_type`.

File: python/ribasim/ribasim/db_utils.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from contextlib import closing
+from pathlib import Path
+from sqlite3 import Connection, connect
+
+
+def esc_id(identifier: str) -> str:
+    """Escape SQLite identifiers."""
+    return '"' + identifier.replace('"', '""') + '"'
```
Comment (deltamarnix) [inline python/ribasim/ribasim/db_utils.py:8]: Would something like this be more readable?

```suggestion
    identifier = identifier.replace('"', '""')
    return f'"{identifier}"'
```

File: python/ribasim/ribasim/db_utils.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    with closing(connect(db_path)) as connection:
+        if exists(connection, "ribasim_metadata"):
+            with closing(connection.cursor()) as cursor:
+                cursor.execute(
+                    "SELECT value FROM ribasim_metadata WHERE key='schema_version'"
+                )
+                result = int(cursor.fetchone()[0])
+        else:
+            result = 0
+    return result
```
Comment (deltamarnix) [inline python/ribasim/ribasim/db_utils.py:53]:
```suggestion
        if not exists(connection, "ribasim_metadata"):
          return 0
        with closing(connection.cursor()) as cursor:
            cursor.execute(
                "SELECT value FROM ribasim_metadata WHERE key='schema_version'"
               )
                return int(cursor.fetchone()[0])
```

File: python/ribasim/ribasim/migrations.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        warnings.warn("Migrating outdated Basin / Static table.", UserWarning)
+        df.drop("urban_runoff", inplace=True, axis=1)
+
+    return df
+
+
+def basintimeschema_migration(df: DataFrame) -> DataFrame:
+    if "urban_runoff" in df.columns:
+        warnings.warn("Migrating outdated Basin / Time table.", UserWarning)
+        df.drop("urban_runoff", inplace=True, axis=1)
```
Comment (deltamarnix) [inline python/ribasim/ribasim/migrations.py:41]: Like codecov is saying, tests would be great.

File: ribasim_qgis/core/geopackage.py [hunk 4]
```diff
@@ -50,6 +51,25 @@ def layers(path: Path) -> list[str]:
     return layers
 
 
+def write_schema_version(path: Path, version: int = 1) -> None:
```
Comment (deltamarnix) [inline ribasim_qgis/core/geopackage.py:54]: I would also love to see tests for this function.

**Review by deltamarnix — CHANGES_REQUESTED** [review]
I want to see tests for the migration and the writing of the version files (both qgis and ribasim python)

File: python/ribasim/ribasim/db_utils.py (hunk 2 above)
Comment (deltamarnix) [reply python/ribasim/ribasim/db_utils.py:53]: It looks like a lot, but it's a simple if-inversion actually.

**Review by deltamarnix — APPROVED** [review]
If @SnippenE agrees that loading an old model in QGIS does not do the migration, but crashes instead, I am okay with the changes so far.

**Pull request merged**